File: code_files_MATLAB/hac_robust.m
function tab_data=hac_robust(reg)


res = reg.Residuals.Raw(~isnan(reg.Residuals.Raw));
% Number of observations
n = length(res);
% compute autocorrelation and confidence bounds
[acf, lags, bounds] = autocorr(res, 'NumLags', n-1);
% Identify significant lags (where autocorrelation exceeds bounds)
significant = (acf(2:end) > bounds(1)) | (acf(2:end) < -bounds(1));  % skip lag 0
% Find the largest significant lag
max_lag = find(significant, 1, 'last'); 
[EstCoeffCov,se,coeff] = hac(reg,display='off',weights="QS",bandwidth=max_lag);
tstat=coeff./se;
pval=2*(1-tcdf(abs(tstat),reg.NumObservations-reg.NumVariables));

num_rows=length(coeff); %number of rows in table
num_cols=6; %number of columns in table
tab=NaN([num_rows num_cols]); %preallocate with NaNs
tab(:,1)=coeff;
tab(:,2)=se;
tab(:,3)=tstat;
tab(:,4)=pval;
tab(1,5)=reg.Rsquared.Adjusted;
tab(1,6)=reg.NumObservations; 

% Define column headers
var_names = {'Coeff', 'SE', 'tStat', 'pVal', 'AdjR2', 'Nobs'};

% Convert to table
tab_data = array2table(tab, 'VariableNames', var_names);

end

File: code_files_R/hac_robust.R
# ================================================================
# FUNCTION: hac_robust(reg)
# PURPOSE:
#   Compute Newey–West HAC (Heteroskedasticity and Autocorrelation
#   Consistent) robust standard errors for a linear model and
#   output a formatted table like the MATLAB version.
# ================================================================
# INPUT:
#   reg : lm() regression object
# ================================================================
# OUTPUT:
#   tab_data : data.frame with columns
#     Coeff | SE | tStat | pVal | AdjR2 | Nobs
# ================================================================

hac_robust <- function(reg) {
  # --- 1. Extract residuals ---
  res <- residuals(reg)
  res <- res[!is.na(res)]
  n <- length(res)
  
  # --- 2. Compute autocorrelation and find max significant lag ---
  acf_obj <- acf(res, plot = FALSE)
  acf_vals <- acf_obj$acf[-1]       # skip lag 0
  bounds <- qnorm((1 + 0.95) / 2) / sqrt(n)
  significant <- (acf_vals > bounds) | (acf_vals < -bounds)
  if (any(significant)) {
    max_lag <- which(significant)[length(which(significant))]
  } else {
    max_lag <- 0
  }
  
  # --- 3. Compute HAC robust covariance matrix using QS weights ---
  # Equivalent to MATLAB hac(reg, weights="QS", bandwidth=max_lag)
  cov_hac <- NeweyWest(reg, lag = max_lag, prewhite = FALSE, adjust = TRUE)
  
  # --- 4. Extract coefficients and statistics ---
  coeff <- coef(reg)
  se <- sqrt(diag(cov_hac))
  tstat <- coeff / se
  df <- reg$df.residual
  pval <- 2 * (1 - pt(abs(tstat), df = df))
  
  # --- 5. Construct summary table (same columns as MATLAB) ---
  num_rows <- length(coeff)
  num_cols <- 6
  tab <- matrix(NA, nrow = num_rows, ncol = num_cols)
  
  tab[, 1] <- coeff
  tab[, 2] <- se
  tab[, 3] <- tstat
  tab[, 4] <- pval
  tab[1, 5] <- summary(reg)$adj.r.squared
  tab[1, 6] <- nobs(reg)
  
  colnames(tab) <- c("Coeff", "SE", "tStat", "pVal", "AdjR2", "Nobs")
  tab_data <- as.data.frame(tab)
  
  # --- 6. Return neatly formatted result ---
  return(tab_data)
}
